### ISSUE TITLE
Can't set headers after they are sent exception - Sails 0.10.5, Socket.io 0.9.17

### ISSUE THREAD

**Issue by sobtiankit** [balderdashy/sails#2859]
Hi

We at Postman are big fans, and users of Sails JS. We are facing an issue in production running Sails 0.10.5, socket.io - 0.9.17 and PM2 in cluster mode deployed across multiple machines. I am using the redis for sockets and sessions. 

Every once a while, the server restarts with the following exception:

```
Error: Can't set headers after they are sent.
sync-prod-10 (err):    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:335:11)
sync-prod-10 (err):    at ServerResponse.res.setHeader (/home/postman/node_modules/sails/node_modules/express/node_modules/connect/lib/patch.js:63:22)
sync-prod-10 (err):    at ServerResponse.<anonymous> (/home/postman/node_modules/sails/node_modules/express/node_modules/connect/lib/patch.js:80:14)
sync-prod-10 (err):    at Array.forEach (native)
sync-prod-10 (err):    at ServerResponse.res.writeHead (/home/postman/node_modules/sails/node_modules/express/node_modules/connect/lib/patch.js:79:28)
sync-prod-10 (err):    at XHRPolling.doWrite (/home/postman/node_modules/sails/node_modules/socket.io/lib/transports/xhr-polling.js:66:17)
sync-prod-10 (err):    at XHRPolling.HTTPPolling.write (/home/postman/node_modules/sails/node_modules/socket.io/lib/transports/http-polling.js:132:8)
sync-prod-10 (err):    at XHRPolling.Transport.onDispatch (/home/postman/node_modules/sails/node_modules/socket.io/lib/transport.js:204:10)
sync-prod-10 (err):    at /home/postman/node_modules/sails/node_modules/socket.io/lib/manager.js:341:17
sync-prod-10 (err):    at RedisClient.message (/home/postman/node_modules/sails/node_modules/socket.io/lib/stores/redis.js:130:24)
sync-prod-10 (err):    at RedisClient.emit (events.js:129:20)
sync-prod-10 (err):    at RedisClient.return_reply (/home/postman/node_modules/sails/node_modules/socket.io/node_modules/redis/index.js:601:22)
sync-prod-10 (err):    at RedisReplyParser.<anonymous> (/home/postman/node_modules/sails/node_modules/socket.io/node_modules/redis/index.js:265:14)
sync-prod-10 (err):    at RedisReplyParser.emit (events.js:107:17)
sync-prod-10 (err):    at RedisReplyParser.add_multi_bulk_reply (/home/postman/node_modules/sails/node_modules/socket.io/node_modules/redis/lib/parser/javascript.js:312:14)
sync-prod-10 (err):    at RedisReplyParser.send_reply (/home/postman/node_modules/sails/node_modules/socket.io/node_modules/redis/lib/parser/javascript.js:273:18)
sync-prod-10 (err):    at RedisReplyParser.execute (/home/postman/node_modules/sails/node_modules/socket.io/node_modules/redis/lib/parser/javascript.js:223:22)
```

I have gone through my code, and I am pretty sure there's nowhere I'm sending a response on the same res object twice. I think I can isolate it to Socket.io, but can anyone confirm that? 

Will an upgrade to Sails 0.11 with Socket.io 1.0+ help? It's a non trivial exercise for us, so I'd prefer to not be forced to upgrade, if there's a workaround.

**Comment by RWOverdijk** [comment]
It looks like you're trying to set a header after sending the response. This doesn't necessarily mean that you're trying to send a response on the same object twice. It can also mean you have a race condition somewhere. You expect some process to be done working, and send the response, while in reality it's not.

**Comment by sobtiankit** [comment]
Hmmm. While given the way the code is organized, I doubt this is the case. Still, I'm going to revisit everything with this in mind. Thanks :) 
Any idea how I can go about debugging this in a production setup? I am unable to reproduce this locally.

As an addon, I saw a fairly similar problem statement in another open ticket: https://github.com/balderdashy/sails/issues/2538

**Comment by tjwebb** [comment]
> While given the way the code is organized, I doubt this is the case

This is definitely the case. You are either responding twice, or you're setting headers after you've sent a response to the client. Can you post a link to your sails project?

**Comment by sobtiankit** [comment]
I finally managed to track this down. I was never setting any headers manually, so following the suggestions above, the only way this could have occurred was by responding on the same res object twice. 

To track it, I set a flag, alreadySent, on the res object in all modules in the responses folder. As I expected, I was not responding twice.

Looking at the stack trace printed above in more detail, the problem was originating from the XHRPolling module of Socket.io. I disabled XHR as a transport, allowing only websockets (which is how it should have been in the first place, actually)  and I haven't seen any crashes since. 

I'm closing this issue, but I am not sure where the problem is, socket.io or sails' implementation of it. @mikermcneil thanks for your help the other day. Tagging you in case you want to look at it further.
